packages/core/src/controllers/W3aCtrl.ts

```diff
@@ -0,0 +1,105 @@
+import { proxy, ref, subscribe as valtioSub } from 'valtio/vanilla'
+import type { W3aCtrlState } from '../types/controllerTypes'
+import { IFRAME_API, IFRAME_ID, IS_SERVER, call, messageHandler } from '../utils/W3aUtil'
+
+// -- initial state ------------------------------------------------ //
+const state = proxy<W3aCtrlState>({
+  iFrameMsgIndex: 0,
+  iFrameReady: false,
+  providerInited: false,
+  iFrame: null as unknown as HTMLIFrameElement,
+  notifyReady: [] as (() => void)[]
+})
+
+// -- controller --------------------------------------------------- //
+export const W3aCtrl = {
+  state,
+
+  subscribe(callback: (newState: W3aCtrlState) => void) {
+    return valtioSub(state, () => callback(state))
+  },
+
+  setIFrame() {
```

Comment:
All of this logic should be taken outside of controller and into web3account sdk. We are doing too much here. Role of controller is to be used by ui and call methods on underlying sdk's / manage state. They should be very simple